File: Python/155.py
# Note:
    # Can keep min in a separate stack and only push to that stack if element < min
    # When popping elements, need to check if pop'd element is also top of min_stack
    # This works because for any element greater than the min, it must be popped before
        # the min, so there will always be a smaller element than it in the stack
    # This is an example of a monotonic stack
class MinStack:

    def __init__(self):
        self.x = []
        self.min_x = []

    def push(self, val: int) -> None:
        self.x.append(val)
        if not self.min_x or val <= self.min_x[-1]: self.min_x.append(val)

    def pop(self) -> None:
        if self.x:
            if self.x[-1] == self.min_x[-1]:
                self.min_x.pop()
            self.x.pop()

    def top(self) -> int:
        if self.x: return self.x[-1]

    def getMin(self) -> int:
        if self.min_x: return self.min_x[-1]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: C++/155.cpp
/*
- Hint: you can always use multiple data structures to make more functional ones
- Notice that simply tracking a minimum doesn't work because once the current min is popped you can't recall priors
- In particular, can use a regular stack for most functions and a monotonic stack for the minimum-tracking
*/
class MinStack {
public:
    MinStack() {
        
    }
    
    void push(int val) {
        stk.push(val);
        if(min_stk.empty() || val < min_stk.top().first){
            min_stk.push({val, 1});
        }
        else if(val == min_stk.top().first){
            min_stk.top().second++;
        }
    }
    
    void pop() {
        if(stk.top() == min_stk.top().first){
            min_stk.top().second--;
        }
        if(min_stk.top().second == 0){
            min_stk.pop();
        }
        stk.pop();
    }
    
    int top() {
        return stk.top();
    }
    
    int getMin() {
        return min_stk.top().first;
    }
private:
    stack<int> stk;
    stack<pair<int, int>> min_stk;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */